### PR TITLE
Added mirror to TTs

### PR DESCRIPTION
**Pull request by henrymkw**

### CODE REVIEW

File: payload/egg/core/eggScene.hh [hunk 1]
```diff
@@ -17,8 +17,24 @@ public:
...
+    }
+
+    Scene *getChild() {
+        return child;
+    }
+
+    u32 getSceneID() {
+        return sceneID;
+    }
+
```
Comment (vabold) [inline payload/egg/core/eggScene.hh:31]: Use const qualifiers on these functions.

File: payload/egg/core/eggSceneManager.cc [hunk 2]
```diff
@@ -76,7 +79,21 @@ void SceneManager::reinitCurrentScene() {
...
+    auto *saveManager = System::SaveManager::Instance();
+    auto setting = saveManager->getSetting<SP::ClientSettings::Setting::TAMirror>();
+    // This is a hack to get mirror TTs working. Restarting a race from mirror to non mirror causes graphical bugs. Opted to reload the entire track as a simple fix.
+    if (rc->raceScenario().mirror && (setting == SP::ClientSettings::TAMirror::Disable)) {
+        Scene *parent = currScene->getParent();
+        u32 sceneID = currScene->getSceneID();
+        destroyScene(currScene);
+        createScene(sceneID, parent);
+    }
+    else {
```
Comment (vabold) [inline payload/egg/core/eggSceneManager.cc:93]: Formatting.

File: payload/game/system/GhostWriter.cc [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#include "GhostWriter.hh"
+#include "RaceConfig.hh"
+#include "InputManager.hh"
+
+extern "C" {
+    #include "revolution.h"
+}
```
Comment (vabold) [inline payload/game/system/GhostWriter.cc:7]: Formatting.

File: payload/game/system/GhostWriter.cc [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            newTrick = System::Trick::Right;
+        }
+        else if (rawTrick == System::Trick::Right) {
+            newTrick = System::Trick::Left;
+        }
+    } 
+    REPLACED(writeFrame)(rawButtons, newStickX, rawStickY, newTrick);
+}
+
+}
```
Comment (vabold) [inline payload/game/system/GhostWriter.cc:34]: Add newline.

File: payload/game/system/GhostWriter.hh [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    public:
+    void REPLACED(writeFrame)(u16 rawButtons, u8 rawStickX, u8 rawStickY, u32 rawTrick);
+    REPLACE void writeFrame(u16 rawButtons, u8 rawStickX, u8 rawStickY, u32 rawTrick);
+
+    private:
+    u8 _00[0x1c - 0x00];
+};
+static_assert(sizeof(GhostWriter) == 0x1c);
+
+}
```
Comment (vabold) [inline payload/game/system/GhostWriter.hh:17]: Add newline.

File: payload/game/system/GhostWriter.hh [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+#include <Common.hh>
+
+namespace System {
+
+class GhostWriter {
+    public:
+    void REPLACED(writeFrame)(u16 rawButtons, u8 rawStickX, u8 rawStickY, u32 rawTrick);
+    REPLACE void writeFrame(u16 rawButtons, u8 rawStickX, u8 rawStickY, u32 rawTrick);
+
+    private:
```
Comment (vabold) [inline payload/game/system/GhostWriter.hh:12]: Formatting.

File: payload/game/system/RaceConfig.cc [hunk 7]
```diff
@@ -112,6 +113,19 @@ void RaceConfig::ConfigurePlayers(Scenario &scenario, u32 screenCount) {
...
 }
 
+void RaceConfig::initRace() {
+    REPLACED(initRace)();
+    auto *saveManager = System::SaveManager::Instance();
+    auto setting = saveManager->getSetting<SP::ClientSettings::Setting::TAMirror>();
+    // Switch the race to mirror if the mirror TT setting is enabled.
+    if (m_raceScenario.gameMode == GameMode::TimeAttack && setting == SP::ClientSettings::TAMirror::Enable) {
+        m_raceScenario.mirror = true;
+    }
```
Comment (vabold) [inline payload/game/system/RaceConfig.cc:123]: Formatting.

**Review by vabold — CHANGES_REQUESTED** [review]
Here is my incredibly useful PR review.

File: payload/egg/core/eggScene.hh (hunk 1 above)
Comment (riidefi) [reply payload/egg/core/eggScene.hh:31]: SP style is more like ```cpp
u32 sceneID() const {
    return m_sceneID;
}```

**Comment by riidefi** [comment]
Nice work!

File: payload/egg/core/eggScene.hh [hunk 8]
```diff
@@ -17,8 +17,24 @@ public:
...
 private:
-    u8 _04[0x30 - 0x04];
+    u8 _04[0x20 - 0x04];
+    Scene *parent;
+    Scene *child;
+    u32 sceneID;
```
Comment (stblr) [inline payload/egg/core/eggScene.hh:36]: `m_`

File: payload/egg/core/eggSceneManager.hh [hunk 9]
```diff
@@ -22,7 +22,9 @@ private:
     static void PushDolphinSpeed(u32 percent);
     static void PopDolphinSpeed();
 
-    u8 _00[0x2c - 0x00];
+    u8 _00[0x0c - 0x00];
+    Scene *currScene;
```
Comment (stblr) [inline payload/egg/core/eggSceneManager.hh:26]: `m_`

File: payload/egg/core/eggSceneManager.cc [hunk 10]
```diff
@@ -9,6 +9,9 @@ extern "C" {
 #include <sp/cs/RaceManager.hh>
 #include <sp/cs/RoomManager.hh>
 
+#include <game/system/RaceConfig.hh>
+#include <game/system/SaveManager.hh>
+
```
Comment (stblr) [inline payload/egg/core/eggSceneManager.cc:14]: https://github.com/stblr/mkw-sp/issues/245#issuecomment-1213490665

File: payload/game/system/GhostWriter.cc (hunk 3 above)
Comment (stblr) [inline payload/game/system/GhostWriter.cc:7]: https://github.com/stblr/mkw-sp/issues/245#issuecomment-1213490665

File: payload/game/system/InputManager.cc [hunk 11]
```diff
@@ -1,4 +1,5 @@
 #include "InputManager.hh"
+#include "RaceConfig.hh"
```
Comment (stblr) [inline payload/game/system/InputManager.cc:2]: https://github.com/stblr/mkw-sp/issues/245#issuecomment-1213490665

File: payload/game/system/RaceConfig.cc [hunk 12]
```diff
@@ -5,6 +5,7 @@ extern "C" {
 #include "game/system/SaveManager.h"
 }
 #include "game/system/SaveManager.hh"
+#include <sp/settings/ClientSettings.hh>
```
Comment (stblr) [inline payload/game/system/RaceConfig.cc:8]: https://github.com/stblr/mkw-sp/issues/245#issuecomment-1213490665

File: payload/game/system/GhostFile.hh [hunk 13]
```diff
@@ -97,6 +97,7 @@ struct SPFooter {
     u32 shroomStrategy : 15;
     bool isVanilla : 1;
     bool hasSimplifiedControls : 1;
+    bool setInMirror : 1;
 
     static constexpr u32 MAGIC = 0x53504744; // SPGD
     static constexpr u32 VERSION = 4;
```
Comment (stblr) [inline payload/game/system/GhostFile.hh:103]: Please bump, and adapt `SPFooter::checkSize`.

**Review by stblr — CHANGES_REQUESTED** [review]

**Review by stblr — APPROVED** [review]

**Pull request merged**